src/realm/sync/binding_callback_thread_observer.hpp

```diff
@@ -20,23 +20,121 @@
-class BindingCallbackThreadObserver {
-public:
+struct BindingCallbackThreadObserver {
+    using NotificationCallback = std::function<void()>;
+    using ErrorCallback = std::function<void(const std::exception&)>;
+
+    // Create a BindingCallbackThreadObserver that can be used in SyncClientConfig
+    BindingCallbackThreadObserver(std::optional<NotificationCallback>&& did_create_thread,
+                                  std::optional<NotificationCallback>&& will_destroy_thread,
+                                  std::optional<ErrorCallback>&& error_handler)
+        : m_create_thread_callback{std::move(did_create_thread)}
+        , m_destroy_thread_callback{std::move(will_destroy_thread)}
+        , m_handle_error_callback{std::move(error_handler)}
+    {
+    }
+
+    virtual ~BindingCallbackThreadObserver() = default;
+
+    // Set the global thread observer with the provided (optional) callback functions
+    static void set_global_thread_observer(std::unique_ptr<BindingCallbackThreadObserver>&& observer_ptr)
+    {
+        if (!observer_ptr) {
+            BindingCallbackThreadObserver::reset_global_thread_observer();
+            return; // early return
+        }
+
+        std::lock_guard<std::mutex> lock{BindingCallbackThreadObserver::m_mutex};
+        BindingCallbackThreadObserver::m_instance.reset(observer_ptr.release());
+    }
+
+    // Returns true if the global binding callback thread observer is set, otherwise false
+    static bool has_global_thread_observer()
+    {
+        return bool(BindingCallbackThreadObserver::m_instance);
+    }
+
+    // Resets the global thread observer so no more callback functions will be called
+    static void reset_global_thread_observer()
+    {
+        std::lock_guard<std::mutex> lock{BindingCallbackThreadObserver::m_mutex};
+        BindingCallbackThreadObserver::m_instance.reset();
+    }
+
+    ///
+    /// Execution Functions - check for a valid instance and if the function was set
+    ///
+
+    // BindingCallbackThreadObserver class that will call will_destroy_thread() when destroyed
+    struct ThreadGuard {
+        ~ThreadGuard()
+        {
+            BindingCallbackThreadObserver::call_will_destroy_thread(m_observer);
+        }
+        // Constructor that only works with the global thread observer
+        ThreadGuard() = default;
+
+        // Constructor that works with either the local or global thread observer
+        ThreadGuard(const std::shared_ptr<BindingCallbackThreadObserver>& observer_ptr)
+            : m_observer{observer_ptr}
+        {
+        }
+
+    private:
+        std::shared_ptr<BindingCallbackThreadObserver> m_observer;
+    };
+
     // This method is called just before the thread is started
-    virtual void did_create_thread() = 0;
+    // This takes an optional reference to an observer_ptr and decides whether to use the passed in or global observer
```

Comment:
I guess the global one is used if nothing is passed. Otherwise the one passed is used. If that's the case the comment is a bit misleading.

Reply:
Updated comments to be more clear